3-pipeline/src/main/scala/riscv/core/fivestage_final/InstructionDecode.scala

```diff
@@ -152,13 +152,25 @@ class InstructionDecode extends Module {
     val clint_jump_address     = Output(UInt(Parameters.AddrWidth)) // clint.io.jump_address
     val if_jump_flag           = Output(Bool())                     // ctrl.io.jump_flag , inst_fetch.io.jump_flag_id
     val if_jump_address        = Output(UInt(Parameters.AddrWidth)) // inst_fetch.io.jump_address_id
+    val uses_rs1_id = Output(Bool()) // tells Control/Forwarding whether rs1 is valid for this instruction
+    val uses_rs2_id = Output(Bool()) // tells Control/Forwarding whether rs2 is valid for this instruction
   })
   val opcode = io.instruction(6, 0)
   val funct3 = io.instruction(14, 12)
   val funct7 = io.instruction(31, 25)
   val rd     = io.instruction(11, 7)
   val rs1    = io.instruction(19, 15)
   val rs2    = io.instruction(24, 20)
+  val usesRs2 = (opcode === InstructionTypes.RM) ||
+              (opcode === InstructionTypes.S)  ||
+              (opcode === InstructionTypes.B)
+
+  val usesRs1 = !(opcode === Instructions.jal) &&
+               !(opcode === Instructions.lui) &&
+               !(opcode === Instructions.auipc)
```

Comment:
Mixed camelCase and snake_case:
```scala
  val usesRs2 = ...  // camelCase                                                                                                                      
  io.uses_rs2_id := usesRs2  // snake_case    
```

Should be consistent - use snake_case per project conventions:                                                                                       
```scala
  val uses_rs2 = ...                                                                                                                                   
  io.uses_rs2_id := uses_rs2
```